pkg/controller/statefulset/stateful_set_utils.go

```diff
@@ -215,6 +215,25 @@ func isCreated(pod *v1.Pod) bool {
 	return pod.Status.Phase != ""
 }
 
+// isStatefullyStuck returns true if a pod in a stateful set is stuck
+// due to a previously bad roll-out. We can detect this by checking all of:
+// 1) The pod is in a pending state
+// 2) The pod is at a different revision than the update revision
+// 3) The update strategy is rolling
+// 4) The set and the pod are at different generations
+func isStatefullyStuck(set *apps.StatefulSet, pod *v1.Pod, currentRevision, updateRevision *apps.ControllerRevision) bool {
+	return isPending(pod) &&
+		getPodRevision(pod) == currentRevision.Name &&
+		currentRevision.Name != updateRevision.Name &&
+		set.Spec.UpdateStrategy.Type == apps.RollingUpdateStatefulSetStrategyType &&
+		set.ObjectMeta.Generation != pod.ObjectMeta.Generation
```

Comment:
Pod's `metadata.generation` is not set. It's `0` and it's not updated even if some of the mutable fields of a `PodSpec` (`spec.containers[*].image`,  `spec.initContainers[*].image`, `spec.activeDeadlineSeconds`) are updated:

https://github.com/kubernetes/kubernetes/blob/4847e7fa8d0e2f62bd807e6bb837c4a7d9740bfc/pkg/registry/core/pod/strategy.go#L64-L82

Reply:
You are totally right @mvladev  - that was my mistake. I dropped the generation off of my check.

Would you happen to have any other ideas for other checks that might go here? I think this check is enough on it's own, but in StatefulSet land, I know there's an abundance of caution.